python/dgl/sparse.py

```diff
@@ -366,124 +366,75 @@ def _bwd_segment_cmp(feat, arg, m):
+        The number of node types for the returned graph (must be either 1 or 2).
 
     Returns
     -------
-    dgl.sparse.CSRMatrix
-        The result
+    C : HeteroGraphIndex
+        The output graph index.
+    C_weights : Tensor
+        The edge weights of the output graph.
     """
-    A_indptr = F.zerocopy_from_numpy(A.indptr)
-    A_indices = F.zerocopy_from_numpy(A.indices)
-    A_data = F.zerocopy_from_numpy(A.data)
-    B_indptr = F.zerocopy_from_numpy(B.indptr)
-    B_indices = F.zerocopy_from_numpy(B.indices)
-    B_data = F.zerocopy_from_numpy(B.data)
-    C_indptr, C_indices, C_data = _CAPI_DGLCSRMM(
-        A.shape[0], A.shape[1], B.shape[1],
-        F.to_dgl_nd(A_indptr),
-        F.to_dgl_nd(A_indices),
-        F.to_dgl_nd(A_data),
-        F.to_dgl_nd(B_indptr),
-        F.to_dgl_nd(B_indices),
-        F.to_dgl_nd(B_data))
-    return CSRMatrix(
-        F.from_dgl_nd(C_data),
-        F.from_dgl_nd(C_indices),
-        F.from_dgl_nd(C_indptr),
-        A.shape[0],
-        B.shape[1])
-
-def csrsum(As):
-    """Sparse-sparse matrix summation.
-
-    This is an internal function whose interface is subject to changes.
+    C, C_weights = _CAPI_DGLCSRMM(
+        A, F.to_dgl_nd(A_weights), B, F.to_dgl_nd(B_weights), num_vtypes)
+    return C, F.from_dgl_nd(C_weights)
+
+def csrsum(As, A_weights):
+    """Return a graph whose adjacency matrix is the sparse matrix summation
+    of the given list of graphs.
 
     Parameters
     ----------
-    As : List[dgl.sparse.CSRMatrix]
-        List of scipy sparse matrices in CSR format.
+    As : list[HeteroGraphIndex]
+        The input graph indices.
+    A_weights : list[Tensor]
+        The edge weights of graph A as 1D tensor.
 
     Returns
     -------
-    dgl.sparse.CSRMatrix
-        The result
+    C : HeteroGraphIndex
+        The output graph index.
+    C_weights : Tensor
+        The edge weights of the output graph.
     """
-    A_indptr = [F.zerocopy_from_numpy(x.indptr) for x in As]
-    A_indices = [F.zerocopy_from_numpy(x.indices) for x in As]
-    A_data = [F.zerocopy_from_numpy(x.data) for x in As]
-    C_indptr, C_indices, C_data = _CAPI_DGLCSRSum(
-        As[0].shape[0], As[0].shape[1],
-        [F.to_dgl_nd(x) for x in A_indptr],
-        [F.to_dgl_nd(x) for x in A_indices],
-        [F.to_dgl_nd(x) for x in A_data])
-    return CSRMatrix(
-        F.from_dgl_nd(C_data),
-        F.from_dgl_nd(C_indices),
-        F.from_dgl_nd(C_indptr),
-        As[0].shape[0], As[0].shape[1])
-
-def csrmask(A, B):
-    """Sparse-sparse matrix masking operation that computes ``A[B != 0]``.
-
-    This is an internal function whose interface is subject to changes.
+    C, C_weights = _CAPI_DGLCSRSum(As, [F.to_dgl_nd(w) for w in A_weights])
+    return C, F.from_dgl_nd(C_weights)
+
+def csrmask(A, A_weights, B):
+    """Return the weights of A at the locations identical to the sparsity pattern
+    of B.
+
+    In scipy notation this is identical to ``A[B != 0]``.
```

Comment:
> In scipy, A[B != 0] does not return a matrix of the same shape of A and B.

And indeed this function returns an array instead of a matrix.

This function together with `csrsum` and `csrmm` are internal functions.

Reply:
But you should not use "identical to". "similar to" would be better as there is still discrepancy.

Reply:
They are indeed identical.
* `A[B != 0]` returns an array, and here `csrmask` indeed returns an array.
* `A[B != 0]` will return zeros for the non-existent entries in A, and here `csrmask` also does that.

Or do you think `csrmask` *sounds* like masking an matrix and therefore should return a matrix?  I can change the function name if you want to.